rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+				inputs[i].Parent.StateElement = se
+			}
+
+			if inputs[i].Parent.LeafIndex == types.UnassignedLeafIndex {
+				continue
+			} else if inputs[i].Parent.LeafIndex >= cru.State.Elements.NumLeaves {
+				return fmt.Errorf("siacoin input %v is not in the correct state", inputs[i].Parent.ID)
+			}
+			cru.UpdateElementProof(&inputs[i].Parent.StateElement)
+		}
+		base = revertedIndex
+	}
+
+	for _, cau := range applied {
+		if cau.State.Index == target {
+			return nil
+		}
+
+		modified := make(map[types.Hash256]types.StateElement)
+		cau.ForEachSiacoinElement(func(sce types.SiacoinElement, created bool, spent bool) {
+			if created {
+				modified[sce.ID] = sce.StateElement
+			}
+		})
+
+		for i := range inputs {
+			if se, ok := modified[inputs[i].Parent.ID]; ok {
+				inputs[i].Parent.StateElement = se
+			}
+
+			if inputs[i].Parent.LeafIndex == types.UnassignedLeafIndex {
+				continue
+			} else if inputs[i].Parent.LeafIndex >= cau.State.Elements.NumLeaves {
+				return fmt.Errorf("siacoin input %v is not in the correct state", inputs[i].Parent.ID)
+			}
+			cau.UpdateElementProof(&inputs[i].Parent.StateElement)
+		}
+		base = cau.State.Index
+	}
+
+	if base != target {
+		return fmt.Errorf("failed to update basis to target %v, current %v", target, base)
+	}
+	return nil
+}
+
+func (s *Server) handleHostStream(stream net.Conn, log *zap.Logger) {
+	defer stream.Close()
+
+	stream.SetDeadline(time.Now().Add(30 * time.Second)) // set an initial timeout
+	rpcStart := time.Now()
+	id, err := rhp4.ReadID(stream)
+	if err != nil {
+		log.Debug("failed to read RPC ID", zap.Error(err))
+		return
+	}
+	log = log.With(zap.Stringer("rpc", id))
+
+	switch id {
+	case rhp4.RPCSettingsID:
+		err = s.handleRPCSettings(stream)
+	case rhp4.RPCAccountBalanceID:
+		err = s.handleRPCAccountBalance(stream)
+	case rhp4.RPCFormContractID:
+		err = s.handleRPCFormContract(stream)
+	case rhp4.RPCFundAccountsID:
+		err = s.handleRPCFundAccounts(stream)
+	case rhp4.RPCLatestRevisionID:
+		err = s.handleRPCLatestRevision(stream)
+	case rhp4.RPCModifySectorsID:
+		err = s.handleRPCModifySectors(stream)
+	case rhp4.RPCReadSectorID:
+		err = s.handleRPCReadSector(stream)
+	case rhp4.RPCRenewContractID:
+		err = s.handleRPCRenewContract(stream)
+	case rhp4.RPCSectorRootsID:
+		err = s.handleRPCSectorRoots(stream)
+	case rhp4.RPCWriteSectorID:
+		err = s.handleRPCWriteSector(stream)
+	case rhp4.RPCVerifySectorID:
+		err = s.handleRPCVerifySector(stream)
+	default:
+		log.Debug("unrecognized RPC", zap.Stringer("rpc", id))
+		rhp4.WriteResponse(stream, &rhp4.RPCError{Code: rhp4.ErrorCodeBadRequest, Description: "unrecognized RPC"})
+		return
+	}
+	if err != nil {
+		var re *rhp4.RPCError
+		if ok := errors.As(err, &re); ok {
+			rhp4.WriteResponse(stream, re)
+			log.Debug("RPC failed", zap.Error(err), zap.Duration("elapsed", time.Since(rpcStart)))
+		} else {
+			rhp4.WriteResponse(stream, rhp4.ErrHostInternalError.(*rhp4.RPCError))
```

Comment:
is the assert necessary here?

Reply:
`rhp4.ErrHostInternalError` is an `error` so no implementation of Object, but we could change that

Reply:
ah right, gotcha. I think I was getting it mixed up with `rhpv2` which has a `WriteResponseErr`.